R/checks.R

```diff
@@ -1033,4 +1033,109 @@ check_invalid_form2.1 <- function(form){
     ) %>%
     mutate(comment="Invalid or missing entries")
   return(problems)
+}
+
+#' Check that all entries in form3.1 are filled with valid values
+#' 
+#' @param form dataframe containing form3.1
+#' @return a dataframe containing problematic entries
+#' 
+#' @import tidyverse
+#' @export
+check_invalid_form3.1 <- function(form) {
+  problems <- form %>% 
+    mutate(
+      is_trans_column_invalid_or_missing = is_invalid_na_or_n(trans) |
+        (trans == 1 &
+           (is_invalid_na_or_n(transsp) | transsp == 0 |
+              is_invalid_or_na(bhelm) | is_invalid_or_na(mhelm) |
+              is_invalid_or_na(tbsbelt) | is_invalid_or_na(asbelt) | 
+              is_invalid_or_na(othtrans) |
+              (transsp == 6 & (is_n(bhelm) | bhelm == 0)) |
+              (transsp == 7 & (is_n(mhelm) | mhelm == 0)) |
+              (transsp == 8 & (is_n(tbsbelt) | tbsbelt == 0)) |
+              (transsp == 9 & (is_n(asbelt) | asbelt == 0)) |
+              (transsp == 10 & is_n(othtrans)) | othtrans == 0)) |
+        (trans == 0 & is_invalid_or_na(transsp)),
+      is_fall_column_invalid_or_missing = is_invalid_na_or_n(fall) | 
+        (fall == 1 & 
+           (is_invalid_na_or_n(fallfrom) | fallfrom == 0 | 
+              is_invalid_or_na(lowhigh) |
+              (fallfrom == 2 & (is_n(lowhigh) | lowhigh == 0)) )) |
+        (fall == 0 & is_invalid_or_na(fallfrom)),
+      is_intent_column_invalid_or_missing = is_invalid_na_or_n(intent) | 
+        (intent == 1 &
+           (is_invalid_na_or_n(intentsp) | intentsp == 0 | 
+              is_invalid_or_na(othinten) |
+              (intentsp == 6 & (is_invalid_na_or_n(othinten) | othinten == 0)) )) |
+        (intent == 0 & is_invalid_or_na(intent)), 
+      is_strklift_column_invalid_or_missing = is_invalid_na_or_n(strklift) | 
+        (strklift == 1 &
+           (is_invalid_na_or_n(stliftsp) | stliftsp == 0 | 
+              is_invalid_or_na(othstlif) |
+              (stliftsp == 4 & (is_invalid_na_or_n(othstlif) | othstlif == 0)) )) |
+        (strklift == 0 & is_invalid_or_na(strklift)),
+      is_othmech_column_invalid_or_missing = is_invalid_na_or_n(othmech) | 
+        (othmech == 1 &
+           (is_invalid_na_or_n(omechsp) | omechsp == 0 | 
+              is_invalid_or_na(othmoth) |
+              (omechsp == 5 & (is_invalid_na_or_n(othmoth) | othmoth == 0)) )) |
+        (othmech == 0 & is_invalid_or_na(othmech)), 
+      num_of_mechanism_of_injury = 
+        (trans != 0) + (fall != 0) + (intent != 0) + (strklift != 0) + (othmech != 0),
```

Comment:
I would suggest using trans == 1 instead, since != 0 could mean that NA, Missing, N, etc. are also filtered.